### DISCUSSION TITLE
How to make RustDesk builds work on Windows 7 (and probably Windows 8)

### DISCUSSION THREAD

**Discussion by Nadahar** [rustdesk/rustdesk#7503] [Q&A]
This isn't a request for support, I've been struggling with this for a couple of days and finally found the culprit. I'm thus writing this merely to share my findings with others. Rust has in [their infinite wisdom](https://github.com/rust-lang/compiler-team/issues/651) (and with good help from Microsoft I'm sure) decided to stop supporting all Windows versions before Windows 10 (for a programming language that is quite serious, and this is a good reason to stay way clear of Rust) after version 1.75.

The result when trying to run an executable compiled with a more recent version of rust is the following error:
```
The procedure entry point ProcessPrng could not be located in the dynamic link library bcryptprimitives.dll.
```
I've dug into the issue a little bit, and this seems to be part of a request to Windows' random generator. Microsoft has done some shuffling of how things are organized in the DLLs, and the change done to Rust has no obvious "benefit" and I strongly suspect deliberate sabotage.

Anyway, the "fix" is sticking to Rust 1.75. The way to do this so that the builds actually work isn't that straight forward though, since the whole toolchain must be held back. What I found is that the `rustup` toolchain must be set to `1.75.0-x86_64-pc-windows-msvc` for `x86_64` builds. If you don't want to specify the toolchain for every call (including modifying the python build scripts), it's easiest to set this as the default with `rustup default 1.75.0-x86_64-pc-windows-msvc`. For the `x86` builds it's even "harder" because it needs a nightly toolchain to support the `x86` calling convention - but I found that using `nightly-2023-12-21-i686-pc-windows-msvc` (which should correspond to 1.75) works.

I have no Windows 8 installations, neither physical nor virtual, so I haven't been able to test Windows 8, but I'm pretty sure that the issue applies to Windows 8 as well - since the intention behind this most likely is to make things stop working for the versions of Windows that Microsoft wants to "kill".

I'm considering to make a PR modifying the build scripts to accommodate this, but it's not trivial for me to test the workflows to see that it actually works so I won't make any promises. The change in Rust has been made after the release of `1.2.3-1`, so the binaries in that release work on Windows 7.

**Comment by rustdesk** [comment]
bcryptprimitives.dll is introduced by new Rust? Nobody report this issue on Rust repo?

**Reply by Nadahar** [reply]
I don't know why, `bcryptprimitives.dll` exists in previous Windows version too I think, but it has something to do with how they call functions or which function they use. Usually applications (or "frameworks", standard libraries) don't call it directly, they call one of several other undocumented API which then again calls `ProcessPrng`. Here is a [discussion on the topic regarding Go](https://github.com/golang/go/issues/33542). These calls aren't documented and is not in "the official API" as far as I understand, but people have found ways to call them that have worked for ages. I don't really understand the details, but as far as I can tell people are committing "optimizations" that might be a few nanoseconds quicker while killing compatibility with "unsupported" systems. I could have misunderstood something though.

**Comment by rustdesk** [comment]
https://github.com/rust-lang/rust/pull/121337/
I will change rust to 1.75 in the CI.

Thank you very much.

**Reply by Nadahar** [reply]
Yeah - you found it (I never managed to find the issue concerning Rust for some reason). But, according to https://github.com/rust-lang/rust/pull/121337, the "old" way should still be used with pre-Windows 10 versions - which sounds like a good solution to me. My problems in recent days indicates that this isn't actually the case, so maybe a bug report could be submitted with Rust if the intention is that this should actually work also on pre-Windows 10 versions.

**Reply by rustdesk** [reply]
https://github.com/rustdesk/rustdesk/commit/28b376d2d29965c8c1c2b6bc970fc853d9d35343
https://github.com/rustdesk/rustdesk/commit/b77add97db0eb39d3924e61e61a7f66f5a1d6fd4
https://github.com/rustdesk/rustdesk/commit/1fa3837bb0d26dfa1f97e935107cc9fb5fe106a7

**Reply by rustdesk** [reply]
@Nadahar @21pages please test if it works on win7

**Reply by Nadahar** [reply]
@rustdesk I can't find a binary attached to the workflow run... am I looking in the wrong place? Or do I have to run the workflow myself?

**Reply by rustdesk** [reply]
Building

**Reply by 21pages** [reply]
Now it works

**Reply by Nadahar** [reply]
I don't know how @21pages built it, but my manual build process is quite different from the workflows so it wouldn't be a fruitful test to use. So, while it's taken me some time, I've now modified the workflows to run in my fork and built the latest master with GitHub. I've tested both the `x86` and the `x86_64` versions on 32 and 64 bits Windows 7, and they both work perfectly fine 👍 

Thank you for prompt action 😄

**Comment by deadash** [comment]
after testing
cargo +nightly-2024-01-25 build --target i686-pc-windows-msvc
Works, later versions fail
Stable version 1.77.0 should work, 1.77.1/2 does not

**Reply by deadash** [reply]
For more information, refer to the PR at: https://github.com/rust-lang/rust/pull/121337

**Reply by rustdesk** [reply]
Interesting, we just tested 1.77.2, it works on windows 7. 1.78 does not.

![image](https://github.com/user-attachments/assets/655b680f-ea64-4b99-b048-b905e1ecc634)

![image](https://github.com/user-attachments/assets/24c2e1d0-9914-47c8-82a9-82c04cfb91d3)

**Comment by asorc** [comment]
Did anyone have any problem with the Visual studio tools? Because I tried to install them, but It's not possible on Windows 7 with the usual installed at least.
